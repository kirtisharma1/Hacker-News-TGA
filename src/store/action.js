import {
  UPDATE_FEED,
  UPDATE_PAGE,
  UPDATE_UPVOTED_LIST,
  UPDATE_HIDE_LIST,
  UPDATE_SORT_TYPE,
} from "./constants";

export const updateFeed = (data, dispatch) => {
  return dispatch({
    type: UPDATE_FEED,
    payload: data,
  });
};

export const updatePage = (data, dispatch) => {
  return dispatch({
    type: UPDATE_PAGE,
    payload: data,
  });
};

export const updateUpvotedList = (data, dispatch) => {
  return dispatch({
    type: UPDATE_UPVOTED_LIST,
    payload: data,
  });
};

export const updateHideList = (data, dispatch) => {
  return dispatch({
    type: UPDATE_HIDE_LIST,
    payload: data,
  });
};

export const updateSortType = (data, dispatch) => {
  return dispatch({
    type: UPDATE_SORT_TYPE,
    payload: data,
  });
};
