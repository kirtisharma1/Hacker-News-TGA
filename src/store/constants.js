export const UPDATE_FEED = "UPDATE_FEED";
export const UPDATE_SORT_TYPE = "UPDATE_SORT_TYPE";
export const UPDATE_UPVOTED_LIST = "UPDATE_UPVOTED_LIST";
export const UPDATE_HIDE_LIST = "UPDATE_HIDE_LIST";
export const UPDATE_PAGE = "UPDATE_PAGE";
export const HEADER_TITLE = "Hacker News";
export const HEADER_TOP_LABEL = "top";
export const HEADER_NEW_LABEL = "new";
export const HIDE_LABEL = "hide";
export const MORE_LABEL = "more";
export const LOADING_TEXT = "Loading...";
