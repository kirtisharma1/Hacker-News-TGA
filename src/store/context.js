import React from "react";

export const GlobalContext = React.createContext();

export const GlobalContextProvider = ({ data, children }) => (
  <GlobalContext.Provider value={data}>{children}</GlobalContext.Provider>
);
