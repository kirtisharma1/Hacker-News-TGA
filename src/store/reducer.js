import {
  UPDATE_FEED,
  UPDATE_PAGE,
  UPDATE_UPVOTED_LIST,
  UPDATE_HIDE_LIST,
  UPDATE_SORT_TYPE,
} from "./constants";
import { HEADER_TOP_LABEL } from "../constants";

export const initialState = {
  feed: [],
  sortType: HEADER_TOP_LABEL,
  upvotedList: [],
  hideList: [],
  page: 0,
};

export const GlobalReducer = (state, action) => {
  switch (action.type) {
    case UPDATE_FEED:
      return {
        ...state,
        feed: [...action.payload],
      };
    case UPDATE_PAGE:
      return {
        ...state,
        page: action.payload,
      };
    case UPDATE_UPVOTED_LIST:
      return {
        ...state,
        upvotedList: [...action.payload],
      };
    case UPDATE_HIDE_LIST:
      return {
        ...state,
        hideList: [...action.payload],
      };
    case UPDATE_SORT_TYPE:
      return {
        ...state,
        sortType: action.payload,
      };
    default:
      return state || initialState;
  }
};
