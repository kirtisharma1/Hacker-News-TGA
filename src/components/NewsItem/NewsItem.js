import React, { useContext } from "react";
import PropTypes from "prop-types";
import { GlobalContext } from "../../store/context";
import { HIDE_LABEL } from "../../constants";
import { updateUpvotedList, updateHideList } from "../../store/action";
import { timeSince } from "../../utils";

export default function NewsItem({ newsItem, order, isUpvoted }) {
  const { state, dispatch } = useContext(GlobalContext);
  const { hideList, upvotedList } = state;
  const getShade = () => {
    switch (true) {
      case newsItem.points > 150:
        return "lightest";
      case newsItem.points > 100:
        return "lighter";
      case newsItem.points > 75:
        return "light";
      case newsItem.points > 50:
        return "normal";
      default:
        return "dark";
    }
  };
  const getDomain = (url) => {
    const arr = url.split("/");
    return arr[2];
  };

  const upvoteNewsItem = (newsItem) => {
    const newList = [...upvotedList];
    const index = newList.findIndex(
      (item) => item.objectID === newsItem.objectID
    );
    if (index >= 0) {
      newList.splice(index, 1);
    } else {
      newList.push(newsItem);
    }

    updateUpvotedList(newList, dispatch);
  };

  const hideNewsItem = (newsItem) => {
    const newList = [...hideList];
    const itemIndex = newList.findIndex(
      (item) => item.objectID === newsItem.objectID
    );
    if (itemIndex >= 0) {
      newList.splice(itemIndex, 1);
    } else {
      newList.push(newsItem);
    }

    updateHideList(newList, dispatch);
  };

  return (
    <article
      className={`news-item  ${order % 2 === 0 ? "even" : "odd"}`}
      data-testid="news-item"
    >
      <span className="news-item__comment-count">{newsItem.num_comments}</span>
      <span className={`news-item__points news-item--${getShade()}`}>
        {newsItem.points}
      </span>
      <button
        type="button"
        className="btn-empty news-item__upvote"
        onClick={() => upvoteNewsItem(newsItem)}
      >
        {isUpvoted}
      </button>
      <p className="news-item__title">
        {newsItem.title}
        <span>
          (
          {newsItem.url && (
            <a target="_blank" rel="noopener noreferrer" href={newsItem.url}>
              {getDomain(newsItem.url)}
            </a>
          )}
          ) by&nbsp;
          <b>{newsItem.author}</b>
          &nbsp;
          <date>{`${timeSince(Date.parse(newsItem.created_at))} ago`}</date>
        </span>
        &nbsp;[
        <button
          type="button"
          className="btn-empty news-item__hide"
          onClick={() => hideNewsItem(newsItem)}
        >
          {HIDE_LABEL}
        </button>
        ]
      </p>
    </article>
  );
}

NewsItem.propTypes = {
  newsItem: PropTypes.shape({
    num_comments: PropTypes.number,
    title: PropTypes.string,
    points: PropTypes.number,
    url: PropTypes.string,
    author: PropTypes.string,
    created_at: PropTypes.string,
  }).isRequired,
  order: PropTypes.number.isRequired,
  isUpvoted: PropTypes.string,
};

NewsItem.defaultProps = {
  newsItem: {
    num_comments: 0,
    title: "",
    points: 0,
    url: "",
    author: "",
    created_at: "",
  },
  order: 0,
  isUpvoted: "",
};
