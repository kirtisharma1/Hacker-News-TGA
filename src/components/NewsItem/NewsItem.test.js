import React from "react";
import { render, fireEvent, wait } from "@testing-library/react";
import NewsItem from "./NewsItem";
import { GlobalContextProvider } from "../../store/context";
import { initialState } from "../../store/reducer";
import { HIDE_LABEL } from "../../constants";
const mockFeed = [
  {
    created_at: "2020-04-11T18:46:08.000Z",
    title: "John Conway has died",
    url: "https://twitter.com/CardColm/status/1249038195880341505",
    author: "ColinWright",
    points: 27,
    story_text: null,
    comment_text: null,
    num_comments: 156,
    story_id: null,
    story_title: null,
    story_url: null,
    parent_id: null,
    created_at_i: 1586630768,
    _tags: ["story", "author_ColinWright", "story_22843306", "front_page"],
    objectID: "22843306",
    _highlightResult: {
      title: {
        value: "John Conway has died",
        matchLevel: "none",
        matchedWords: [],
      },
      url: {
        value: "https://twitter.com/CardColm/status/1249038195880341505",
        matchLevel: "none",
        matchedWords: [],
      },
      author: {
        value: "ColinWright",
        matchLevel: "none",
        matchedWords: [],
      },
    },
  },
  {
    created_at: "2020-04-11T18:46:08.000Z",
    title: "John Conway has died",
    url: "https://twitter.com/CardColm/status/1249038195880341505",
    author: "ColinWright",
    points: 110,
    story_text: null,
    comment_text: null,
    num_comments: 16,
    story_id: null,
    story_title: null,
    story_url: null,
    parent_id: null,
    created_at_i: 1586630768,
    _tags: ["story", "author_ColinWright", "story_22843306", "front_page"],
    objectID: "22843306",
    _highlightResult: {
      title: {
        value: "John Conway has died",
        matchLevel: "none",
        matchedWords: [],
      },
      url: {
        value: "https://twitter.com/CardColm/status/1249038195880341505",
        matchLevel: "none",
        matchedWords: [],
      },
      author: {
        value: "ColinWright",
        matchLevel: "none",
        matchedWords: [],
      },
    },
  },
  {
    created_at: "2020-04-11T18:46:08.000Z",
    title: "John Conway has died",
    url: "https://twitter.com/CardColm/status/1249038195880341505",
    author: "ColinWright",
    points: 90,
    story_text: null,
    comment_text: null,
    num_comments: 156,
    story_id: null,
    story_title: null,
    story_url: null,
    parent_id: null,
    created_at_i: 1586630768,
    _tags: ["story", "author_ColinWright", "story_22843306", "front_page"],
    objectID: "22843306",
    _highlightResult: {
      title: {
        value: "John Conway has died",
        matchLevel: "none",
        matchedWords: [],
      },
      url: {
        value: "https://twitter.com/CardColm/status/1249038195880341505",
        matchLevel: "none",
        matchedWords: [],
      },
      author: {
        value: "ColinWright",
        matchLevel: "none",
        matchedWords: [],
      },
    },
  },
  {
    created_at: "2020-04-11T18:46:08.000Z",
    title: "John Conway has died",
    url: "https://twitter.com/CardColm/status/1249038195880341505",
    author: "ColinWright",
    points: 70,
    story_text: null,
    comment_text: null,
    num_comments: 156,
    story_id: null,
    story_title: null,
    story_url: null,
    parent_id: null,
    created_at_i: 1586630768,
    _tags: ["story", "author_ColinWright", "story_22843306", "front_page"],
    objectID: "22843306",
    _highlightResult: {
      title: {
        value: "John Conway has died",
        matchLevel: "none",
        matchedWords: [],
      },
      url: {
        value: "https://twitter.com/CardColm/status/1249038195880341505",
        matchLevel: "none",
        matchedWords: [],
      },
      author: {
        value: "ColinWright",
        matchLevel: "none",
        matchedWords: [],
      },
    },
  },
  {
    created_at: "2020-04-12T02:22:08.000Z",
    title: "Nasty macOS flaw is bricking MacBooks: Don't install this update",
    url:
      "https://www.tomsguide.com/news/nasty-macos-flaw-is-bricking-macbooks-dont-install-this-update",
    author: "Corrado",
    points: 531,
    story_text: null,
    comment_text: null,
    num_comments: 355,
    story_id: null,
    story_title: null,
    story_url: null,
    parent_id: null,
    created_at_i: 1586658128,
    _tags: ["story", "author_Corrado", "story_22845629", "front_page"],
    objectID: "22845629",
    _highlightResult: {
      title: {
        value:
          "Nasty macOS flaw is bricking MacBooks: Don't install this update",
        matchLevel: "none",
        matchedWords: [],
      },
      url: {
        value:
          "https://www.tomsguide.com/news/nasty-macos-flaw-is-bricking-macbooks-dont-install-this-update",
        matchLevel: "none",
        matchedWords: [],
      },
      author: { value: "Corrado", matchLevel: "none", matchedWords: [] },
    },
  },
];
const mockItem = {
  created_at: "2020-04-11T18:46:08.000Z",
  title: "John Conway has died",
  url: "https://twitter.com/CardColm/status/1249038195880341505",
  author: "ColinWright",
  points: 2037,
  story_text: null,
  comment_text: null,
  num_comments: 156,
  story_id: null,
  story_title: null,
  story_url: null,
  parent_id: null,
  created_at_i: 1586630768,
  _tags: ["story", "author_ColinWright", "story_22843306", "front_page"],
  objectID: "22843306",
  _highlightResult: {
    title: {
      value: "John Conway has died",
      matchLevel: "none",
      matchedWords: [],
    },
    url: {
      value: "https://twitter.com/CardColm/status/1249038195880341505",
      matchLevel: "none",
      matchedWords: [],
    },
    author: {
      value: "ColinWright",
      matchLevel: "none",
      matchedWords: [],
    },
  },
};

const dispatch = jest.fn();
describe("NewsItem component", () => {
  let result;
  beforeEach(() => {
    result = render(
      <GlobalContextProvider
        data={{
          state: { ...initialState, feed: mockFeed, upvotedList: mockFeed },
          dispatch,
        }}
      >
        <NewsItem newsItem={mockItem} order={2} isUpvoted={"\u25BC"} />
      </GlobalContextProvider>
    );
  });
  test("snapshot", () => {
    expect(result.asFragment()).toMatchSnapshot();
  });

  test("upvote", () => {
    fireEvent.click(result.getByText("\u25BC"));
    wait(() => {
      expect(result.queryByText("\u25B2")).toBeInTheDocument();
      fireEvent.click(result.getByText("\u25B2"));
    });

    wait(() => {
      expect(result.getByText("\u25BC")).toBeInTheDocument();
    });
  });

  test("hide", () => {
    fireEvent.click(result.getByText(HIDE_LABEL));
  });

  test("shades", () => {
    fireEvent.click(result.getByText(HIDE_LABEL));
  });
});
