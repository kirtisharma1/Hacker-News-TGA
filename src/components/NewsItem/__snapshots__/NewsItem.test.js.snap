// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewsItem component snapshot 1`] = `
<DocumentFragment>
  <article
    class="news-item  odd"
    data-testid="news-item"
  >
    <span
      class="news-item__comment-count"
    >
      156
    </span>
    <span
      class="news-item__points news-item--lightest"
    >
      2037
    </span>
    <button
      class="btn-empty news-item__upvote"
      type="button"
    >
      ▼
    </button>
    <p
      class="news-item__title"
    >
      John Conway has died
      <span>
        (
        <a
          href="https://twitter.com/CardColm/status/1249038195880341505"
          rel="noopener noreferrer"
          target="_blank"
        >
          twitter.com
        </a>
        ) by 
        <b>
          ColinWright
        </b>
         
        <date>
          1 day ago
        </date>
      </span>
       [
      <button
        class="btn-empty news-item__hide"
        type="button"
      >
        hide
      </button>
      ]
    </p>
  </article>
</DocumentFragment>
`;
